### DISCUSSION TITLE
Compute derivative from fitted model

### DISCUSSION THREAD

**Discussion by llamm-de** [MilesCranmer/PySR#414] [Q&A] [answered]
Dear @MilesCranmer & community,
first of all, thanks for developing the wonderful SR package. It is a pleasure working with it.

Once I have trained my model using the MLJ interface, I would not only like to use ```predict(mach, X)``` but also compute the derivative of my model expression with respect to ```X```. I know that there exists the ```eval_diff_tree_array()``` function, but I am not sure on how to call this function outside e.g. a custom loss function. I would somehow need to get access to the ```tree``` of the best fit from my model to pass it to ```eval_diff_tree_array()```. Is this possible? Or does anyone have a better idea on how to tackle the issue?

Thanks a lot in advance.

Bests
Lukas

**Comment by MilesCranmer** [comment, marked as answer]
Hi @llamm-de,

Thanks for the message. This is a bit tricky as MLJ does not have an interface for this, so you will have to use the low-level SymbolicRegression.jl calls directly. The easiest way to do this right now would be as follows:

```julia
using MLJ
using SymbolicRegression as SR

...
model = ...
mach = ...
...

# Get the tree:
r = report(mach)
tree = r.equations[r.best_idx]

# Get the transposed X:
X_t, _, _ = SR.MLJInterfaceModule.get_matrix_and_info(X, model.dimensions_type)

# Get options from mach:
options = mach.fitresult.options

# Use:
eval_grad_tree_array(tree, X_t, options; variable=true/false)  # or eval_diff_tree_array
```

But it would definitely nice if there was a function to pass this through automatically.

Cheers,
Miles

**Comment by llamm-de** [comment]
Hey @MilesCranmer,

thanks for the quick answer. It works like a charm. :blush: :+1: 

Bests Lukas
